napari/utils/_injection/_type_resolution.py

```diff
@@ -0,0 +1,251 @@
+from __future__ import annotations
+
+import sys
+import types
+import typing
+from functools import lru_cache
+from inspect import Signature
+from typing import TYPE_CHECKING, Any, Callable, Dict, Optional, Tuple
+
+if TYPE_CHECKING:
+    from typing import _get_type_hints_obj_allowed_types
+
+PY39_OR_GREATER = sys.version_info >= (3, 9)
+
+
+@lru_cache(maxsize=1)
+def _napari_names() -> Dict[str, Any]:
+    """Napari names to inject into local namespace when evaluating type hints."""
+    import napari
+    from napari import components, layers, viewer
+
+    return {
+        'napari': napari,
+        **viewer.__dict__,
+        **layers.__dict__,
+        **components.__dict__,
+    }
+
+
+@lru_cache(maxsize=1)
+def _typing_names():
+    return {**typing.__dict__, **types.__dict__}  # noqa: TYP006
+
+
+def resolve_type_hints(
+    obj: _get_type_hints_obj_allowed_types,
+    globalns: Optional[dict] = None,
+    localns: Optional[dict] = None,
+    include_extras: bool = False,
+    inject_napari_namespace: bool = True,
+) -> Dict[str, Any]:
+    """Return type hints for an object.
+
+    This is a small wrapper around `typing.get_type_hints()` that adds napari
+    namespaces to the global and local namespaces.
+
+    see docstring for :func:`typing.get_type_hints`.
+
+    Parameters
+    ----------
+    obj : module, class, method, or function
+        must be a module, class, method, or function.
+    globalns : Optional[dict]
+        optional global namespace, by default None.
+    localns : Optional[dict]
+        optional local namespace, by default None.
+    include_extras : bool
+        If `False` (the default), recursively replaces all 'Annotated[T, ...]'
+        with 'T'.
+    inject_napari_namespace : bool
+        If `True` (the default), will inject all names from
+        `napari.components`, `napari.layers`, and `napari.viewer`
+        into the namespace when evaluating types.
+
+    Returns
+    -------
+    Dict[str, Any]
+        mapping of object name to type hint for all annotated attributes of `obj`.
+
+    Examples
+    --------
+    >>> def func(image: 'Image', layer_list: 'LayerList') -> 'Labels': ...
+    >>> resolve_type_hints(func)
+    {
+        'image': <class 'napari.layers.image.image.Image'>,
+        'layer_list': <class 'napari.components.layerlist.LayerList'>,
+        'return': <class 'napari.layers.labels.labels.Labels'>
+    }
+    """
+    _localns = dict(_typing_names())
+    if inject_napari_namespace:
+        _localns.update(_napari_names())
+    if localns:
+        _localns.update(localns)  # explicitly provided locals take precedence
+    return typing.get_type_hints(
+        obj, globalns=globalns, localns=_localns, include_extras=include_extras
```

Comment:
`include_extras` is 3.9+ only. and I believe is the source of tests failures.
... we could drop 3.8, but I believe this is a bit early to do so.

Reply:
i put the fix in the wrong function :joy: will move